06.Lesson_ThinkingProperties/ThinkingPropertiesExercise/MeetingScheduler/meetingscheduler.spec

```diff
@@ -0,0 +1,208 @@
+    uint256 stateBefore = getStartTimeById(meetingId);
+    // we want to ensure we're in the right initial state, which if
+    // we're trying to transition to ENDED means we must be in the STARTED
+    // (status = 2) state
+    require stateBefore == 2;
+
+    endMeeting(e, meetingId);
+
+    uint256 endTimeAfter = getEndTimeById(meetingId);
+    uint8 stateAfter = getStateById(meetingId);
+
+    assert (stateAfter == 3, "unexpected state");
+    assert (e.block.timestamp > endTimeAfter, "block.timestamp is before the end time");
+}
+
+rule stateOfEndedParametric(uint256 meetingId) {
+    env e;
+    method f;
+    calldataarg args;
+
+    require getStateById(meetingId) != 3;
+
+    f(e, args);
+    uint8 stateAfter = getStateById(meetingId);
+
+    // ensure that the status transitioned to ENDED
+    require stateAfter == 3;
+
+    uint256 endTimeAfter = getEndTimeById(meetingId);
+
+    assert (e.block.timestamp >= endTimeAfter, "block.timestamp is before the end time");
+}
+
+// once a meeting is in the ENDED state, every non-view function called
+// on it should revert (ENDED should be a terminal state)
+rule endStateShouldCauseFailureForAllNonViewFunctions(uint256 meetingId, uint256 startTime, uint256 endTime) {
+    method f;
+
+    require f.isView == false;
+
+    uint8 stateBefore = getStateById(meetingId);
+    require stateBefore == 3;
+
+    callFunctionWithParams(f, meetingId, startTime, endTime);
+
+    assert false;
+}
+
+rule uninitializedStateTransition(uint256 meetingId) {
+    method f;
+    env e;
+    calldataarg args;
+
+    uint256 stateBefore = getStateById(meetingId);
+
+    require stateBefore == 0;
+
+    f(e, args);
+
+    uint256 stateAfter = getStateById(meetingId);
+
+    assert (stateBefore != stateAfter => f.selector == scheduleMeeting(uint256,uint256,uint256).selector &&
+        stateAfter == 1, "the state transition is invalid");
+}
+
+// TODO write rules 6, 7, 8, 9 from 06.Lesson_ThinkingProperties/meetingscheduler_properties_solution, which
+// all are of the same form as the above rule
+
+// numParticipants should never decrease
+rule numParticipantsShouldNeverDecrease(method f, uint256 meetingId) {
+    env e;
+    calldataarg args;
+
+    uint256 numParticipantsBefore = getNumOfParticipents(meetingId);
+    uint8 stateBefore = getStateById(meetingId);
+
+    // need to handle the case where scheduleMeeting can be called (and set
+    // numParticipants to 0) and not called in a separate manner
+    if (stateBefore == 0) {
+        // if the state is unitialized, then we assume nothing has been done with
+        // the meeting, so assuming numParticipants == 0 is a sound assumption
+        require numParticipantsBefore == 0;
+    }
+
+    f(e, args);
+
+    uint256 numParticipantsAfter = getNumOfParticipents(meetingId);
+
+    assert (numParticipantsBefore <= numParticipantsAfter, "numParticipants decreased");
+}
+
+// TODO: figure out why the assert at the end is passing, even though it shouldn't.
```

Comment:
@tadams can you tell why this rule passes? Please read the comment to see why I think it should not be passing.

Reply:
lastReverted is updated after every function call, so here it will be true only if getStateById reverted. You can store lastReverted in a variable before calling getStateById to get around this.

Reply:
done, thanks!